File: genkey.py
from bip_utils import Bip39MnemonicGenerator, Bip39SeedGenerator, Bip44, Bip44Coins
from pygments import highlight, lexers, formatters
import sys

def generate_address(prefix, coin_type, mnemonic):
    # Generate seed from mnemonic
    seed_bytes = Bip39SeedGenerator(mnemonic).Generate()

    # Create a BIP44 object with the specified coin type
    bip44_mst_ctx = Bip44.FromSeed(seed_bytes, Bip44Coins(coin_type))
    
    # Derive the BIP44 account keys for account 0
    bip44_acc_ctx = bip44_mst_ctx.Purpose().Coin().Account(0)

    # Derive the BIP44 chain keys for external chain
    bip44_chg_ctx = bip44_acc_ctx.Change(Bip44Changes.CHAIN_EXT)

    # Derive the address at index 0
    bip44_addr_ctx = bip44_chg_ctx.AddressIndex(0)

    # Get the private key, public key, and address
    priv_key = bip44_addr_ctx.PrivateKey().Raw().ToHex()
    pub_key = bip44_addr_ctx.PublicKey().ToBech32Addr(prefix)
    address = bip44_addr_ctx.PublicKey().ToAddress()
    return address, pub_key, priv_key

# Main function
def main():
    prefix = input("Enter the prefix (e.g., 'cosmos', 'chihuahua'): ")
    coin_type = int(input("Enter the HD wallet cointype (e.g., 118 for Cosmos): "))
    mnemonic = input("Enter your seed phrase: ")

    try:
        address, pub_key, priv_key = generate_address(prefix, coin_type, mnemonic)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        return

    # Display results
    colored_pubkey = highlight(pub_key, lexers.JsonLexer(), formatters.TerminalFormatter())
    colored_privkey = highlight(priv_key, lexers.TextLexer(), formatters.TerminalFormatter())

    print("\033[91m")  # Red color start
    print(f"Address: {address}\nPublic Key: {colored_pubkey}\nType: local")
    print("\033[0m")  # Reset color

    print("\033[93m")  # Yellow color start
    print(f"Private Key: {colored_privkey}")
    print("*NEVER SHARE THIS OR STORE DIGITALLY*")
    print("\033[0m")  # Reset color

if __name__ == "__main__":
    main()

File: genkey.js
const secp256k1 = require('secp256k1');
const createHash = require('create-hash');
const bip39 = require('bip39');
const { bech32 } = require('bech32');
const readline = require('readline');

function derivePrivateKey(mnemonic) {
    // Generate a seed from the mnemonic
    const seed = bip39.mnemonicToSeedSync(mnemonic);
    // Hash the seed to derive a private key
    const hash = createHash('sha256').update(seed).digest();
    let privateKey = hash;
    // Ensure the private key is valid
    while (!secp256k1.privateKeyVerify(privateKey)) {
        privateKey = createHash('sha256').update(privateKey).digest();
    }
    return privateKey;
}

function createCosmosAddress(privateKey, prefix) {
    // Derive the public key
    const publicKey = secp256k1.publicKeyCreate(privateKey, true);
    // Hash the public key to get the address
    const sha256Hash = createHash('sha256').update(publicKey).digest();
    const ripemd160Hash = createHash('ripemd160').update(sha256Hash).digest();
    // Encode the address using Bech32
    const address = bech32.encode(prefix, bech32.toWords(ripemd160Hash));
    return address;
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question('Enter your mnemonic phrase: ', (mnemonic) => {
    rl.question("Enter the address prefix (e.g., cosmos, chihuahua): ", (prefix) => {
        const privateKey = derivePrivateKey(mnemonic);
        const address = createCosmosAddress(privateKey, prefix);

        console.log(`Address: ${address}`);
        console.log(`Private Key: ${privateKey.toString('hex')}`);

        rl.close();
    });
});
